File: algorithm/go/firstround/40.go
package firstround

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	if candidates == nil || len(candidates) < 1 {
		return nil
	}

	res := make([][]int, 0)
	track := make([]int, 0)

	sort.Ints(candidates)
	backtrack(candidates, track, &res, target, 0, 0)
	return res
}

func backtrack(candidates, track []int, res *[][]int, target, sum, start int) {
	if sum == target {
		tmp := make([]int, len(track))
		copy(tmp, track)
		*res = append(*res, tmp)
		return
	}

	for i := start; i < len(candidates); i++ {
		// 剪枝, 避免重复答案
		if i > start && candidates[i] == candidates[i-1] {
			continue
		}

		sum += candidates[i]
		// 剪枝, 提前返回
		if sum > target {
			return
		}
		track = append(track, candidates[i])

		backtrack(candidates, track, res, target, sum, i+1)

		sum -= candidates[i]
		track = track[:len(track)-1]
	}
}

File: algorithm/cpp/firstround/40.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target);

private:
    void recur(vector<int>& candidates, int target, int startIndex, int sum);
    vector<int> path;
    vector<vector<int>> res;
};

vector<vector<int>> Solution::combinationSum2(vector<int>& candidates, int target) {
    if(candidates.empty()) return {};

    sort(candidates.begin(), candidates.end());
    if(target<candidates.front()) return {};
    recur(candidates, target, 0, 0);
    return res;
}

void Solution::recur(vector<int>& candidates, int target, int startIndex, int sum)
{
    if(sum==target)
    {
        res.emplace_back(path);
        return;
    }

    for(int i=startIndex; i<candidates.size() && (sum+candidates[i]<=target); i++)
    {
        if(i>startIndex && candidates[i]==candidates[i-1]) continue;
        path.emplace_back(candidates[i]);
        recur(candidates, target, i+1, sum+candidates[i]);
        path.pop_back();
    }
    return;
}